// ng2-if-media/dist/lib/if-media.directive.js
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { IfMediaService } from './if-media.service';
var IfMediaDirective = (function () {
    function IfMediaDirective(templateRef, viewContainer, mediaService) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.mediaService = mediaService;
        this.hidden = true;
    }
    Object.defineProperty(IfMediaDirective.prototype, "ngIfMedia", {
        set: function (query) {
            this.media = query;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IfMediaDirective.prototype, "ifMedia", {
        set: function (query) {
            this.media = query;
        },
        enumerable: true,
        configurable: true
    });
    IfMediaDirective.prototype.ngOnInit = function () {
        this.mediaService.registerElement(this, this.media);
    };
    IfMediaDirective.prototype.hide = function () {
        if (!this.hidden) {
            this.viewContainer.clear();
            this.hidden = true;
        }
    };
    IfMediaDirective.prototype.show = function () {
        if (this.hidden) {
            this.viewContainer.createEmbeddedView(this.templateRef);
            this.hidden = false;
        }
    };
    IfMediaDirective.prototype.ngOnDestroy = function () {
        this.mediaService.deregisterElement(this);
    };
    IfMediaDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ifMedia], [ngIfMedia]'
                },] },
    ];
    /** @nocollapse */
    IfMediaDirective.ctorParameters = function () { return [
        { type: TemplateRef, },
        { type: ViewContainerRef, },
        { type: IfMediaService, },
    ]; };
    IfMediaDirective.propDecorators = {
        "ngIfMedia": [{ type: Input },],
        "ifMedia": [{ type: Input },],
    };
    return IfMediaDirective;
}());
export { IfMediaDirective };
